onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc

```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
+  /// </summary>
+  /// <param name="dq_to_look_past">The DQToLookPast to steal the node from</param>
+  /// <returns>The dq_node unique_ptr moved from dq_to_look_past</returns>
+  static std::unique_ptr<api::NodeRef> TakeDQNode(std::optional<DQToLookPast>& dq_to_look_past) {
+    std::unique_ptr<api::NodeRef> node;
+
+    if (dq_to_look_past) {
+      node = std::move(dq_to_look_past->dq_node_);
+      dq_to_look_past = std::nullopt;
+    }
+
+    return node;
+  }
+
+ private:
+  std::unique_ptr<api::NodeRef> dq_node_;
+  QuantizationInfo quant_info_;
+};
+
 /// <summary>
 /// Return a DequantizeLinear node if it's input is a constant initializer and it has a single consumer.
 /// In this case the initializer can be updated in-place by UnsqueezeInput or TransposeInput.
 /// </summary>
 /// <param name="graph">Current graph</param>
 /// <param name="value_name">Value to check if produced by a DQ node who's input is a constant initializer</param>
-/// <returns>NodeRef for DQ node if it meets the requirements.</returns>
-static std::unique_ptr<api::NodeRef> GetDQWithConstInitializerInputAndSingleConsumer(const api::GraphRef& graph,
-                                                                                     std::string_view value_name) {
-  std::unique_ptr<api::NodeRef> result;
+/// <returns>DQToLookPast for DQ node if it meets the requirements, or std::nullopt otherwise.</returns>
+static std::optional<DQToLookPast> GetDQWithConstInitializerInputAndSingleConsumer(const api::GraphRef& graph,
+                                                                                   std::string_view value_name) {
+  std::optional<DQToLookPast> result;
+  std::optional<QuantizationInfo> quant_info;
   auto dq_node = GetDQIfProducingValue(graph, value_name);
 
   if (dq_node) {
     do {
-      auto dq_input = dq_node->Inputs()[0];
+      auto dq_inputs = dq_node->Inputs();
+      auto dq_input = dq_inputs[0];
       auto dq_constant = graph.GetConstant(dq_input);
 
       // input to DQ must be a constant initializer
       if (!dq_constant) {
         break;
       }
 
-      // For now keep it simple and don't support per-axis quantization as that would require updating the axis of
-      // the DQ node during TransposeInputImpl and UnsqueezeInput.
-      auto dq_scale = graph.GetConstant(dq_node->Inputs()[1]);
-      if (!dq_scale || dq_scale->NumElements() != 1) {
+      // Scale input to DQ must be a constant initializer.
+      std::string_view dq_scale_input = dq_inputs[1];
+      auto dq_scale_constant = graph.GetConstant(dq_scale_input);
+      if (!dq_scale_constant) {
+        break;
+      }
+
+      // Zero-point input (if present) must be a constant initializer.
+      if (dq_inputs.size() > 2) {
+        std::string_view dq_zp_input = dq_inputs[2];
+        auto dq_zp_constant = graph.GetConstant(dq_zp_input);
+        if (!dq_zp_constant) {
+          break;
+        }
+      }
```

Comment:
Do these have to be constant initializers or do we just need to know the rank?

Reply:
These had to be constant initializers because we tried to permute all DQ inputs for quantized roi/sizes/scales/pads. However, now that we're not handling DQs for this special case, we only need to be able to know the rank of the scales input. I removed this code because the call to `GetQuantizationInfo` checks that we can get the scale input's rank.